functions/Restore-DbaBackup.ps1

```diff
@@ -0,0 +1,215 @@
+        [parameter(Mandatory = $true,ParameterSetName="Files")]
+		[Alias("ServerInstance", "SqlInstance")]
+		[object]$SqlServer,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[System.Management.Automation.PSCredential]$SqlCredential,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[string]$DatabaseName,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+        [String]$RestoreLocation,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+        [DateTime]$RestoreTime = (Get-Date).addyears(1),
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]  
+		[switch]$NoRecovery,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[switch]$ReplaceDatabase,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[switch]$Scripts,
+        [Parameter(ParameterSetName="Paths")]
+        [Switch]$XpDirTree,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+        [switch]$ScriptOnly,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[switch]$VerifyOnly,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+        [switch]$OlaStyle,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[string]$LogicalFilePrefix,
+        [Parameter(ParameterSetName="Paths")][Parameter(ParameterSetName="Files")]
+		[hashtable]$LogicalFileMapping			
+	)
+    BEGIN
+    {
+        $FunctionName = "Restore-DbaBackup"
+        $BackupFiles = @()
+    }
+    PROCESS
+    {
+        if ($PSCmdlet.ParameterSetName -eq "Paths")
+        {
+            Write-Verbose "$FunctionName : Paths passed in" 
+            foreach ($p in $path)
+            {  
+                if ($XpDirTree)
+                {
+                    $BackupFiles += Get-XPDirTreeRestoreFile -path $p -SqlServer $SqlServer -SqlCredential $SqlCredential
+                }
+                elseif ((Get-Item $p).PSIsContainer -ne $true)
+                {
+                    Write-Verbose "$FunctionName : Single file"
+                    $BackupFiles += Get-item $p
+                } 
+                elseif ($OlaStyle)
+                {
+                    Write-Verbose "$FunctionName : Ola Style"
+                    $BackupFiles += Get-OlaHRestoreFile -path $p
+                } 
+                else 
+                {
+                    Write-Verbose "$FunctionName : Standard Directory"
+                    $BackupFiles += Get-DirectoryRestoreFile -path $p
+                }
+            }
+        }elseif($PSCmdlet.ParameterSetName -eq "Files")
+        {
+            Write-Verbose "$FunctionName : Files passed in $($Files.count)" 
+            Foreach ($File in $Files)
+            {
+                $BackupFiles += $File
```

Comment:
shouldn't there be some validation and error handling here?
passed in $files can contain anything.